File: src/compiler/dateTimePattern.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { checkDateTimePattern } from "./dateTimePattern.ts";

test("valid patterns report nothing", () => {
  const r = checkDateTimePattern("yyyy-MM-dd HH:mm:ss");
  expect(r.invalidLetters).toEqual([]);
  expect(r.footguns).toEqual([]);
});

test("unknown pattern letters are reported", () => {
  expect(checkDateTimePattern("yyyy-jj").invalidLetters).toContain("j");
});

test("quoted literals are ignored", () => {
  const r = checkDateTimePattern("yyyy 'at' HH'h'");
  expect(r.invalidLetters).toEqual([]); // the letters in 'at'/'h' are literal
});

test("Y without a week field is flagged as a footgun", () => {
  const r = checkDateTimePattern("YYYY-MM-dd");
  expect(r.footguns.map(f => f.letter)).toContain("Y");
  // ...but Y with a week field is legitimate
  expect(checkDateTimePattern("YYYY-'W'ww").footguns).toEqual([]);
});

test("D alongside a month is flagged (day-of-year vs day-of-month)", () => {
  expect(checkDateTimePattern("yyyy-MM-DD").footguns.map(f => f.letter)).toContain("D");
});

test("h without am/pm is flagged (12h vs 24h)", () => {
  expect(checkDateTimePattern("hh:mm").footguns.map(f => f.letter)).toContain("h");
  expect(checkDateTimePattern("hh:mm a").footguns).toEqual([]);
});

File: togo/internal/compiler/datetimepattern_test.go
// Port of src/compiler/dateTimePattern.test.ts
package compiler

import (
	"slices"
	"testing"
)

func footgunLetters(r DateTimePatternReport) []string {
	var out []string
	for _, f := range r.Footguns {
		out = append(out, f.Letter)
	}
	return out
}

func TestDateTimeValid(t *testing.T) {
	r := CheckDateTimePattern("yyyy-MM-dd HH:mm:ss")
	if len(r.InvalidLetters) != 0 || len(r.Footguns) != 0 {
		t.Errorf("got %+v, want clean", r)
	}
}

func TestDateTimeUnknownLetter(t *testing.T) {
	if !slices.Contains(CheckDateTimePattern("yyyy-jj").InvalidLetters, "j") {
		t.Error("want 'j' flagged")
	}
}

func TestDateTimeQuoted(t *testing.T) {
	if got := CheckDateTimePattern("yyyy 'at' HH'h'").InvalidLetters; len(got) != 0 {
		t.Errorf("quoted letters flagged: %v", got)
	}
}

func TestDateTimeFootguns(t *testing.T) {
	if !slices.Contains(footgunLetters(CheckDateTimePattern("YYYY-MM-dd")), "Y") {
		t.Error("want Y footgun")
	}
	if got := CheckDateTimePattern("YYYY-'W'ww").Footguns; len(got) != 0 {
		t.Errorf("Y with week field flagged: %v", got)
	}
	if !slices.Contains(footgunLetters(CheckDateTimePattern("yyyy-MM-DD")), "D") {
		t.Error("want D footgun")
	}
	if !slices.Contains(footgunLetters(CheckDateTimePattern("hh:mm")), "h") {
		t.Error("want h footgun")
	}
	if got := CheckDateTimePattern("hh:mm a").Footguns; len(got) != 0 {
		t.Errorf("h with am/pm flagged: %v", got)
	}
}
